tools/pipelines/templates/include-publish-npm-package-steps.yml

```diff
@@ -70,6 +70,16 @@ steps:
       --retry $maximumRetryIfNetworkError \
       --verbose
 
+      if [[ "$(release)" == "release" && ${{ parameters.feedKind }} == "internal-build" ]]; then
+        flub promote package \
+        --version "$(SetVersion.version)" \
+        --orderFile $orderFile \
+        --token ":$ADO_API_TOKEN" \
+        --verbose
+      fi
```

Comment:
Could you take in the `release` and `feedKind` params as flags and then execute the if logic in the flub command? This is OK, but I don't see any advantage to the logic living here instead of in the command.

Reply:
I disagree here. It feels weird that the only use of those flags inside the command is "if you pass value X for this flag, I will exit immediately instead of doing anything". If the command used the values for something else, I would agree with passing them as flags, but as is I think the check is better done outside and only calling `flub promote package` when we want it to actually do its thing.

Reply:
The problem with if/else logic in bash is that few people know the right syntax. Is this correct as written? How do we know? When someone adds a new condition, how will we know it's right? If a newline is added between the conditions for readability, will it break? Is the semicolon before `then` required? It _should_ be simple, but even something like if/else is very easy to get wrong in a shell script, not to mention it's not testable. At least when all the logic is in the command it can be tested exactly as it will run in CI - including the conditions under which it will be skipped.

Is there a way for the command to output something that ADO will pick up and display in UI or next to a "skipped" status or something?

Reply:
For logging, using [these](https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#formatting-commands) could work.

I don't disagree with the disadvantages of bash logic, but I don't love making the CLI weird to accommodate that.

What does everyone think about using an ADO `condition` on the promotion step? Something like (I _think_ the syntax is right but merits testing):

```yaml
- task: Bash@3
  displayName: Promote Packages for ${{ parameters.artifactPath }}
  condition: and(succeeded(), eq(variables.release, "release"), eq(${{ parameters.feedKind }}, "internal-build"))
  env:
    ADO_API_TOKEN: $(System.AccessToken)
  inputs:
    targetType: 'inline'
    workingDirectory: $(Pipeline.Workspace)/pack/${{ parameters.artifactPath }}
    script: |
      set -eux -o pipefail
      echo "Artifact path: ${{ parameters.artifactPath }}"
      orderFile=$(Pipeline.Workspace)/pack/packagePublishOrder-${{ parameters.feedKind }}.txt
      flub promote package \
      --version "$(SetVersion.version)" \
      --orderFile $orderFile \
      --verbose
```

The condition is still pipeline-level logic, so not particularly testable, but I think it's a good compromise.

I think it also has the advantage that maybe at some point we will want to promote a particular prerelease build to the release view in ADO feeds, by running the CLI locally.

Reply:
> What does everyone think about using an ADO condition on the promotion step?

I prefer that to bash, but it still seems like a lot of code. Up to you, Rishhi.